File: src/org/nextras/orm/intellij/annotator/ModifierHighlighterAnnotator.java
package org.nextras.orm.intellij.annotator;

import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.Annotator;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.tree.LeafPsiElement;
import com.intellij.psi.tree.IElementType;
import com.jetbrains.php.lang.documentation.phpdoc.lexer.PhpDocTokenTypes;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocType;
import com.jetbrains.php.lang.documentation.phpdoc.psi.impl.PhpDocTypeImpl;
import org.jetbrains.annotations.NotNull;
import org.nextras.orm.intellij.annotator.highlighter.ModifierHighlighter;
import org.nextras.orm.intellij.parser.PhpDocTagModifier;
import org.nextras.orm.intellij.parser.PhpDocTagModifierClassType;
import org.nextras.orm.intellij.parser.PhpDocTagModifierName;
import org.nextras.orm.intellij.parser.PhpDocTagModifierIdentifier;


public class ModifierHighlighterAnnotator implements Annotator
{
	@Override
	public void annotate(@NotNull PsiElement psiElement, @NotNull AnnotationHolder annotationHolder)
	{
		PsiElement parent = psiElement.getParent();
		IElementType type = psiElement instanceof LeafPsiElement ? psiElement.getNode().getElementType() : null;

		// LBRACE || RBRACE
		if (parent instanceof PhpDocTagModifier && (type == PhpDocTokenTypes.DOC_LBRACE || type == PhpDocTokenTypes.DOC_RBRACE)) {
			annotationHolder.createInfoAnnotation(psiElement, null).setTextAttributes(ModifierHighlighter.BRACES);

		} else if (psiElement instanceof PhpDocTagModifierName) {
			annotationHolder.createInfoAnnotation(psiElement, null).setTextAttributes(ModifierHighlighter.MODIFIER);

		} else if (psiElement instanceof PhpDocTagModifierIdentifier || psiElement instanceof PhpDocTagModifierClassType) {
			annotationHolder.createInfoAnnotation(psiElement, null).setTextAttributes(ModifierHighlighter.IDENTIFIER);
		}
	}
}

File: src/org/nextras/orm/intellij/annotator/ModifierHighlighterAnnotator.kt
package org.nextras.orm.intellij.annotator

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiWhiteSpace
import com.jetbrains.php.lang.documentation.phpdoc.lexer.PhpDocTokenTypes
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocPropertyTag
import com.jetbrains.php.lang.psi.elements.impl.PhpPsiElementImpl
import org.nextras.orm.intellij.annotator.highlighter.ModifierHighlighter

/**
 * This is a modifier highlighter.
 */
class ModifierHighlighterAnnotator : Annotator {
	override fun annotate(psiElement: PsiElement, annotationHolder: AnnotationHolder) {
		// currently PhpDoc property comment is always wrapped in this specific type
		if (psiElement::class != PhpPsiElementImpl::class) return
		if (psiElement.parent !is PhpDocPropertyTag) return

		var isModifier = false
		var isModifierName = false
		var element: PsiElement? = psiElement.node.firstChildNode as? PsiElement ?: return

		while (element != null) {
			if (element.node.elementType == PhpDocTokenTypes.DOC_LBRACE) {
				isModifier = true
				isModifierName = true
				annotationHolder
					.newSilentAnnotation(HighlightSeverity.INFORMATION)
					.range(element)
					.textAttributes(ModifierHighlighter.BRACES)
					.create()

			} else if (element.node.elementType == PhpDocTokenTypes.DOC_RBRACE) {
				isModifier = false
				annotationHolder
					.newSilentAnnotation(HighlightSeverity.INFORMATION)
					.range(element)
					.textAttributes(ModifierHighlighter.BRACES)
					.create()

			} else if (isModifier) {
				if (isModifierName && element !is PsiWhiteSpace) {
					annotationHolder
						.newSilentAnnotation(HighlightSeverity.INFORMATION)
						.range(element)
						.textAttributes(ModifierHighlighter.MODIFIER)
						.create()

				} else if (!isModifierName && (element.node.elementType == PhpDocTokenTypes.DOC_IDENTIFIER || element.node.elementType == PhpDocTokenTypes.DOC_VARIABLE)) {
					annotationHolder
						.newSilentAnnotation(HighlightSeverity.INFORMATION)
						.range(element)
						.textAttributes(ModifierHighlighter.IDENTIFIER)
						.create()

				} else {
					isModifierName = false
				}
			}
			element = element.nextSibling
		}
	}
}
